### ISSUE TITLE
Moving away from Travis

### ISSUE THREAD

**Issue by ziotom78** [litebird/litebird_sim#75]
Latest news: https://blog.travis-ci.com/2020-11-02-travis-ci-new-billing

TLDR: Travis cannot be used any longer for public repositories. We're receiving 1000 "free trial" minutes, and then we either pay for Travis, or we move to other solutions. (Also, they are going to charge Mac OS X support a lot more on, but this is reasonable, as they're forced to use costly hardware to run builds for that platform.)

I see that in the last few months many organisations have moved from Travis to GitHub Actions; other options are Circle CI, Drone CI. I have no strong preference, as I never used anything else than Travis. Do you have any good/bad experience to share about other CI systems?

**Comment by mreineck** [comment]
Github Actions looks attractive, but is currently plagued by nasty security holes (see, e.g. https://bugs.chromium.org/p/project-zero/issues/detail?id=2070&can=2&q=&colspec=ID%20Type%20Status%20Priority%20Milestone%20Owner%20Summary&cells=ids). We might have to wait until the dust settles.

**Comment by ziotom78** [comment]
After Mathew's comment, I investigated a bit the issue with Travis. As explained in [their blog post](https://blog.travis-ci.com/2020-11-02-travis-ci-new-billing), Travis' free plan grants us 10k credits. This is how build time is charged:

- 1 minute on a Linux machine costs 10 credits
- 1 minute on a Mac OS X machine costs 50 credits

At the moment, a Travis build takes 5 minutes on a Linux machine and 6 minutes on a Mac OS X machine, and we have the following builds running for every commit:

| # | OS | Python | MPI |
|---|----|-------|----|
| 1 | Linux | 3.6 | No |
| 2 | Linux | 3.6 | OpenMPI |
| 3 | Linux | 3.6 | Mpich |
| 4 | Linux | 3.7 | No |
| 5 | Linux | 3.7 | OpenMPI |
| 6 | Linux | 3.7 | Mpich |
| 7 | Mac OS X | 3.7 | No |
| 8 | Mac OS X | 3.7 | OpenMPI |
| 9 | Mac OS X | 3.7 | Mpich |

In [PR#74](https://github.com/litebird/litebird_sim/pull/74) I changed builds 4, 5, and 6 to use Python 3.8, and I removed builds 8 and 9 because of some weird problem with the machines provided by Travis, but this latter option was only temporary, and the PR has not been merged yet. Let's suppose that we merge the PR: since it implements caching for Linux machines, this would bring the build time down to 2,5 minutes.

I calculated that launching a CI job means that we take 2,5×10 credits per build on Linux, for a total of 25×6=150 credits, plus 1×50 credits per build on Mac OS X (I count only one build because I adopt the conservative assumption that we give up about testing MPI on Mac OS X). The overall amount of credits per job (Linux+Mac) is thus 150+50=200 credits.

If we do not find a way to speed up builds dramatically, we're going to empty our credit vault in 20 commits. (And to speed up builds we need to run many debug builds on Travis, thus exhausting all our credits in the process!) Given that PR#74 took ~30 commits to reach its present stage, it's clear that we must adopt an alternative solution to Travis ASAP.

**Comment by ziotom78** [comment]
One of the possibilities we discussed was to find money to pay TravisCI, but after some investigations I am even more convinced that this would not be the best option. There have been more and more complaints on technical forums about the way Travis CI has evolved since Idera's acquisition in 2019, and many customers are unhappy with the new management policies:

- https://www.jeffgeerling.com/blog/2020/travis-cis-new-pricing-plan-threw-wrench-my-open-source-works
- https://ropensci.org/technotes/2020/11/19/moving-away-travis/
- https://news.ycombinator.com/item?id=25003387
- https://news.ycombinator.com/item?id=24964601

I am more and more convinced that TravisCI is at a dead end and is no longer worth to invest time on it, even if we might be able to find the money to pay them for keeping running our `.travis.yml` file. It would be far better to move to another solution ASAP, as we might deplete the infamous "10k credits" very quickly.

What are your thoughts?

**Comment by dpole** [comment]
I agree that Travis is too expensive for what we need. I briefly considered to keep applying for the 10k credits until I had a look at Github Actions more closely: moving to GitHub Actions seem the best solution to me. Beyond the specific bug that @mreineck pointed out (which got marked as fixed in the meanwhile), I don't see significant limitations. Note that all our target OS are supported and we would get for free more concurrent jobs than what we currently get on TravisCI (see e.g. https://docs.github.com/en/free-pro-team@latest/actions/reference/usage-limits-billing-and-administration). Having also windows support in the same CI makes is easier to maintain: we could drop AppVeyor in addition to Travis

**Comment by ziotom78** [comment]
PR #77 has moved the CI scripts to GitHub Actions.
